tests/ui/needless_range_loop2.rs

```diff
@@ -117,3 +117,17 @@ mod issue2277 {
         }
     }
 }
+
+mod issue11399 {
+    pub struct Block {
+        data: [i32; 3],
+    }
+
+    pub fn example() {
+        let values: [i8; 3] = [1, 2, 3];
+        let mut block = Block { data: [0; 3] };
+        for i in 0..values.len() {
+            block.data[i] = values[i] as i32;
+        }
```

Comment:
This could be written as

```rust
for (i, value) in values.iter().enumerate() {
     block.data[i] = value as i32;
}
```

Which is more idiomatic IMO and I think the lint already suggests to do. 

Even better might be to suggest using `zip`, but that might change semantics without length information. So I wouldn't go that route.

Reply:
Yeah, they should be treated the same. Now I'm wondering why `data[i]` is not linted :thinking: 

But in this case, I think the fix is incomplete. What about a case like this:

```rust
for i in 0..values.len() {
    some_func()[i] = values[i] as i32;
}
```

Or any other expression that returns something that can be indexed and assigned to. Is the lint triggering for those as well?

Reply:
It looks that the indirect indexing means indexing any expression containing the index pat-var, like `[i + 1]`, other than `[i]`.

Reply:
Ah thanks for the clarification! Then I think there are 2 proper ways to fix this:

1. Make it lint less by just searching for `ExprKind::Index` and check if `i` is used in the index expression and if it is used more than once, don't lint
2. Make it lint more by doing the same check, but then suggest to use `enumerate`.

I think for this PR option 1 is the better / quicker one to fix the issue. Option 2 is more a long term thing.